### PR TITLE
Steven/ribbon treasuries/strike selection

### PR DESCRIPTION
**Pull request by stevenwal**
This is to add a new method for strike selection using a multiplier instead of a targeted delta.

### CODE REVIEW

File: contracts/interfaces/IRibbon.sol [hunk 1]
```diff
@@ -8,6 +8,12 @@ interface IStrikeSelection {
         returns (uint256, uint256);
 
     function delta() external view returns (uint256);
+
+    function getSimpleStrikePrice(
+        uint256 expiryTimestamp,
+        bool isPut,
+        uint256 strikeMultiplier
+    ) external view returns (uint256, uint256);
```
Comment (kenchangh) [inline contracts/interfaces/IRibbon.sol:16]: We should just stick to the `getStrikePrice` function because it shares a common interface with the delta based strike selection

File: contracts/utils/StrikeSelection.sol [hunk 2]
```diff
@@ -152,6 +153,49 @@ contract StrikeSelection is Ownable {
...
+    /**
+     * @notice Gets the strike price by multiplying the current spot price with a multiplier
+     * @param expiryTimestamp is the unix timestamp of expiration
+     * @param isPut is whether option is put or call
+     * @param strikeMultiplier is the multiplier to compute the strike price
+     * @return newStrikePrice is the strike price of the option (ex: for BTC might be 45000 * 10 ** 8)
+     * @return newDelta is the delta of the option given its parameters
+     */
+
+    function getSimpleStrikePrice(
```
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:165]: Instead of writing a new function, we should just create a new contract that uses the new methodology. Call it `PercentStrikeSelection.sol` and the current one `DeltaStrikeSelection.sol`. This helps us switch between the multiple strike selection methodologies

File: contracts/utils/StrikeSelection.sol [hunk 3]
```diff
@@ -152,6 +153,49 @@ contract StrikeSelection is Ownable {
...
+        uint256 annualizedVol =
+            volatilityOracle.annualizedVol(optionsPremiumPricer.pool()).mul(
+                10**10
+            );
+
+        newStrikePrice = isPut
+            ? strikePrice.sub(strikePrice % step)
+            : strikePrice.add(step - (strikePrice % step));
+
+        newDelta = optionsPremiumPricer.getOptionDelta(
```
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:191]: Since the delta doesn't matter for the caller, I think we can comment and say we just set the `newDelta` to 0. This helps us save gas from doing `getOptionDelta`

File: contracts/utils/StrikeSelection.sol [hunk 4]
```diff
@@ -152,6 +153,49 @@ contract StrikeSelection is Ownable {
...
+
+        // asset's annualized volatility
+        uint256 annualizedVol =
+            volatilityOracle.annualizedVol(optionsPremiumPricer.pool()).mul(
+                10**10
+            );
+
+        newStrikePrice = isPut
+            ? strikePrice.sub(strikePrice % step)
+            : strikePrice.add(step - (strikePrice % step));
```
Comment (kenchangh) [inline contracts/utils/StrikeSelection.sol:189]: `strikePrice.add(step.sub(strikePrice % step))` for consistency

File: contracts/interfaces/IRibbon.sol [hunk 5]
```diff
@@ -10,13 +10,29 @@ interface IStrikeSelection {
...
 
+interface IPercentStrikeSelection {
+    function getStrikePrice(
+        uint256 expiryTimestamp,
+        bool isPut,
+        uint256 strikeMultiplier
+    ) external view returns (uint256, uint256);
+
+    function delta() external view returns (uint256);
+}
```
Comment (kenchangh) [inline contracts/interfaces/IRibbon.sol:21]: They should all be united under one interface `IStrikeSelection` rather than being split out. This gives us the ability to be agnostic on what the `getStrikePrice` actually does. We can move the `strikeMultiplier` as a variable on the strike selection contract

File: contracts/libraries/VaultLifecycle.sol [hunk 6]
```diff
@@ -5,7 +5,7 @@ import {SafeMath} from "@openzeppelin/contracts/utils/math/SafeMath.sol";
 import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import {Vault} from "./Vault.sol";
 import {ShareMath} from "./ShareMath.sol";
-import {IStrikeSelection} from "../interfaces/IRibbon.sol";
+import {IDeltaStrikeSelection} from "../interfaces/IRibbon.sol";
```
Comment (kenchangh) [inline contracts/libraries/VaultLifecycle.sol:8]: We can stick to IStrikeSelection https://github.com/ribbon-finance/ribbon-v2/pull/187/files#r770235566

File: contracts/storage/RibbonTreasuryVaultStorage.sol [hunk 7]
```diff
@@ -70,6 +67,12 @@ abstract contract RibbonTreasuryVaultStorageV1 {
     /// @notice Amount locked for scheduled withdrawals last week;
     uint256 public lastQueuedWithdrawAmount;
 
+    /// @notice Period between each options sale.
+    uint256 public period;
+
+    /// @notice Multiplier for strike selection
+    uint256 public strikeMultiplier;
```
Comment (kenchangh) [inline contracts/storage/RibbonTreasuryVaultStorage.sol:74]: This variable should be moved to the `StrikeSelection` contract

File: contracts/utils/PercentStrikeSelection.sol [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+    // step in absolute terms at which we will increment
+    // (ex: 100 * 10 ** assetOracleDecimals means we will move at increments of 100 points)
+    uint256 public step;
+
+    // multiplier to shift asset prices
+    uint256 private immutable assetOracleMultiplier;
+
+    // Delta are in 4 decimal places. 1 * 10**4 = 1 delta.
+    uint256 private constant DELTA_MULTIPLIER = 10**4;
```
Comment (kenchangh) [inline contracts/utils/PercentStrikeSelection.sol:36]: Can remove this because it is not related

File: contracts/utils/PercentStrikeSelection.sol [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+     * @notice Sets new delta value
+     * @param newDelta is the new delta value
+     */
+    function setDelta(uint256 newDelta) external onlyOwner {
+        require(newDelta > 0, "!newDelta");
+        require(newDelta <= DELTA_MULTIPLIER, "newDelta cannot be more than 1");
+        uint256 oldDelta = delta;
+        delta = newDelta;
+        emit DeltaSet(oldDelta, newDelta, msg.sender);
+    }
```
Comment (kenchangh) [inline contracts/utils/PercentStrikeSelection.sol:170]: Can remove this because it is not related

File: contracts/utils/PercentStrikeSelection.sol [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+        uint256 finalDelta,
+        uint256 prevDelta,
+        uint256 strike,
+        bool isPut
+    ) private view returns (uint256) {
+        if (finalDelta != prevDelta) {
+            return strike;
+        }
+        return isPut ? strike.add(step) : strike.sub(step);
+    }
```
Comment (kenchangh) [inline contracts/utils/PercentStrikeSelection.sol:158]: Not used anywhere so we can remove this

File: contracts/utils/PercentStrikeSelection.sol [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+    /**
+     * @notice Rounds to best delta value
+     * @param prevDelta is the delta of the previous strike price
+     * @param currDelta is delta of the current strike price
+     * @param targetDelta is the delta we are targeting
+     * @param isPut is whether its a put
+     * @return the best delta value
+     */
+    function _getBestDelta(
```
Comment (kenchangh) [inline contracts/utils/PercentStrikeSelection.sol:113]: Not used anywhere so we can remove this

File: contracts/utils/PercentStrikeSelection.sol [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    // multiplier to shift asset prices
+    uint256 private immutable assetOracleMultiplier;
+
+    // Delta are in 4 decimal places. 1 * 10**4 = 1 delta.
+    uint256 private constant DELTA_MULTIPLIER = 10**4;
+
+    // ChainLink's USD Price oracles return results in 8 decimal places
+    uint256 private constant ORACLE_PRICE_MULTIPLIER = 10**8;
+
+    event DeltaSet(uint256 oldDelta, uint256 newDelta, address indexed owner);
```
Comment (kenchangh) [inline contracts/utils/PercentStrikeSelection.sol:41]: Can remove this event

File: contracts/utils/PercentStrikeSelection.sol [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+
+    /**
+     * Immutables
+     */
+    IOptionsPremiumPricer public immutable optionsPremiumPricer;
+
+    IVolatilityOracle public immutable volatilityOracle;
+
+    // delta for options strike price selection. 1 is 10000 (10**4)
+    uint256 public delta;
```
Comment (kenchangh) [inline contracts/utils/PercentStrikeSelection.sol:26]: Can remove this because it is not related

File: contracts/utils/PercentStrikeSelection.sol [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+contract PercentStrikeSelection is Ownable {
+    using SafeMath for uint256;
+
+    /**
+     * Immutables
+     */
+    IOptionsPremiumPricer public immutable optionsPremiumPricer;
+
+    IVolatilityOracle public immutable volatilityOracle;
```
Comment (kenchangh) [inline contracts/utils/PercentStrikeSelection.sol:23]: Can remove this because it is not related

File: contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol [hunk 15]
```diff
@@ -77,7 +73,7 @@ contract RibbonTreasuryVault is
     address public immutable OTOKEN_FACTORY;
 
     // The minimum duration for an option auction.
-    uint256 private constant MIN_AUCTION_DURATION = 1 hours;
+    uint256 private constant MIN_AUCTION_DURATION = 15 minutes;
```
Comment (kenchangh) [inline contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol:76]: Set this to 5 minutes so we could potentially go lower

File: contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol [hunk 16]
```diff
@@ -369,9 +314,20 @@ contract RibbonTreasuryVault is
...
+
+    /**
+     * @notice Internal function to convert cap into uint104
+     * @param newCap is the cap to convert into uint104
+     */
+    function _getCap104(uint256 newCap) internal pure returns (uint104) {
         ShareMath.assertUint104(newCap);
-        emit CapSet(vaultParams.cap, newCap);
-        vaultParams.cap = uint104(newCap);
+        return uint104(newCap);
```
Comment (kenchangh) [inline contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol:330]: Could just refactor this functionality into `ShareMath` as the function `ShareMath.toUint104`. Does not have to be specific to `cap`. See how we do it for https://github.com/ribbon-finance/governance/blob/main/contracts/rbn-staking/IncentivizedVotingLockup.sol#L835

File: contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol [hunk 17]
```diff
@@ -443,6 +399,22 @@ contract RibbonTreasuryVault is
         lastStrikeOverrideRound = vaultState.round;
     }
 
+    /**
+     * @notice Set the multiplier for setting the strike price
+     * @param newStrikeMultiplier is the strike multiplier (decimals = 2)
+     */
+    function setStrikeMultiplier(uint256 newStrikeMultiplier)
```
Comment (kenchangh) [inline contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol:406]: Move this down to StrikeSelection

File: contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol (hunk 16 above)
Comment (stevenwal) [reply contracts/vaults/TreasuryVault/RibbonTreasuryVault.sol]: Reverting this to the initial version.

```
function setCap(uint256 newCap) external onlyOwner {
      require(newCap > 0, "!newCap");
      ShareMath.assertUint104(newCap);
      emit CapSet(vaultParams.cap, newCap);
      vaultParams.cap = uint104(newCap);
  }
```


This change was reflected here because it was not merged with the fee-distribution branch yet.

**Review by kenchangh — APPROVED** [review]
🚢
